### ISSUE TITLE
Create a JAX issue for axis dependent jit output bug

### ISSUE THREAD

**Issue by chrisfinlay** [epfl-radio-astro/tabascal#46]
The following example highlights a bug that was found in JAX that only presented itself in specific versions: `jax==0.5.0` - `jax==0.8.1`. As of `jax==0.8.2` (til `jax==0.9.0.1` so far) it appears to be fixed.

The example script used to diagnose the bug existence is

```python
import numpy as np

import jax.numpy as jnp
from jax import jit

X = np.arange(4).reshape(2, 2)
A = np.arange(4).reshape(2, 2)

I = np.array([0, 0, 1, 1])

# Pure Numpy implementation

def f1(x, a):
    return x[:,I] * np.square(a[:,I])

def f1_alt(x, a):
    x_ = np.swapaxes(x, 0, 1)
    a_ = np.swapaxes(a, 0, 1)
    return x_[I] * np.square(a_[I])

def f2(x):
    return np.sum(f1(x, A))

def f2_alt(x):
    return np.sum(f1_alt(x, A))

# Various JAX versions

def g1(x, a):
    return x[:,I] * jnp.square(a[:,I])

jitted_g1 = jit(g1)

def g1_alt(x, a):
    x_ = jnp.swapaxes(x, 0, 1)
    a_ = jnp.swapaxes(a, 0, 1)
    return x_[I] * jnp.square(a_[I])

jitted_g1_alt = jit(g1_alt)

def g2(x):
    return jnp.sum(g1(x, A))

def g2_inner_jit(x):
    return jnp.sum(jitted_g1(x, A))

def g2_inner_jit_alt(x):
    return jnp.sum(jitted_g1_alt(x, A))


print("Pure Numpy")
print(f2(X))

print("Pure Numpy with swapped axes")
print(f2_alt(X))

print("JAX Unjitted")
print(g2(X))

print("JAX inner jitted")
print(g2_inner_jit(X))

print("JAX inner and outer jitted")
print(jit(g2_inner_jit)(X))

print("JAX inner and outer jitted with axes swap")
print(jit(g2_inner_jit_alt)(X))
```

The failed output is:
```
Pure Numpy
72
Pure Numpy with swapped axes
72
JAX Unjitted
72
JAX inner jitted
72
JAX inner and outer jitted
54
JAX inner and outer jitted with axes swap
72
```
where each output should be equal to 72.
